# -*- mode: Snakemake -*-
#
# list the all input files for each step

####################
## qc
####################

### Quality-control reads
#TARGET_QC = expand(str(QC_FP/'paired'/'{sample}_{rp}_fastqc.zip'), sample = Samples.keys(), rp = ['R1','R2'])
TARGET_QC = expand(str(QC_FP/'paired'/'{sample}_{rp}.fastq'), sample = Samples.keys(), rp = ['R1','R2'])

### Remove host reads
TARGET_DECONTAM = expand(str(QC_FP/'decontam'/'{sample}_{rp}.fastq'), sample = Samples.keys(), rp = ['R1', 'R2'])

####################
## classify
####################

### Classify all reads
TARGET_KRAKEN = [str(CLASSIFY_FP/'kraken'/'all_samples.tsv')]
TARGET_CLASSIFY = TARGET_KRAKEN

####################
## assembly
####################

### Assemble contigs
TARGET_ASSEMBLY = (
    expand(str(ASSEMBLY_FP/'{sample}_assembly'/'final-contigs.fa'),
           sample = Samples.keys()) +
    expand(str(ASSEMBLY_FP/'{sample}_assembly'/'.save'),
           sample = Samples.keys()))
## Pair reads
TARGET_PAIR = expand(str(ASSEMBLY_FP/'paired'/'{sample}.assembled.fastq'), sample=Samples.keys())

####################
## annotation
####################

## Annotate all contigs
TARGET_ANNOTATE = expand(str(ANNOTATION_FP/'summary'/'{sample}.tsv'), sample=Samples.keys())

####################
## mapping
####################

## KEGG pathway finder
TARGET_PATH = expand(str(MAPPING_FP/'summary'/'pathfinder'/'{sample}_summary.json'), sample=Samples.keys())

#TARGET_ALL = TARGET_CLASSIFY + TARGET_ASSEMBLY + TARGET_ANNOTATE
TARGET_ALL = TARGET_DECONTAM
